web/src/components/import/DataSheetView.tsx

```diff
@@ -0,0 +1,638 @@
+import Box from '@mui/material/Box';
+import Button from '@mui/material/Button';
+import Typography from '@mui/material/Typography';
+import CloudDownloadIcon from '@mui/icons-material/CloudDownload';
+import CloudUploadIcon from '@mui/icons-material/CloudUpload';
+import PlaylistAddCheckOutlinedIcon from '@mui/icons-material/PlaylistAddCheckOutlined';
+import UndoIcon from '@mui/icons-material/Undo';
+import ResetIcon from '@mui/icons-material/LayersClear';
+import { AgGridReact } from 'ag-grid-react';
+import { useParams } from 'react-router-dom';
+import { getDataJob, submitIngest, updateRows, validateJob } from '../../api/api';
+import { AppConstants, extendedMeasurements, measurements } from '../../common/constants';
+import LoadingOverlay from '../overlays/LoadingOverlay';
+import AlertDialog from '../ui/AlertDialog';
+import FindReplacePanel from './panel/FindReplacePanel';
+import ValidationPanel from './panel/ValidationPanel';
+import eh from './DataSheetEventHandlers';
+import { Paper } from '@mui/material';
+import ReportProblemIcon from '@mui/icons-material/ReportProblem';
+import BackButton from '../ui/BackButton';
+import DataRectificationHandler from './DataRectificationHandler';
+import { CellStyle, CellStyleFunc, ColDef, GridApi, SuppressKeyboardEventParams } from 'ag-grid-enterprise';
+import { AxiosResponse } from 'axios';
+import { ValidationResponse, InternalContext, RowUpdate, StagedRow, JobResponse, SourceJobType, StatusJobType, ExtRow, Measurement, SurveyValidationErrorLevel } from '../../common/types';
+import { CellEditingStoppedEvent, CellValueChangedEvent, FilterChangedEvent, GetContextMenuItemsParams, GridReadyEvent, MenuItemDef, PasteEndEvent, RowDataUpdatedEvent } from 'ag-grid-community';
+
+// |context| is where all custom properties and helper functions
+// associated with the ag-grid are stored
+// see: https://www.ag-grid.com/javascript-grid/context/
+
+const context: InternalContext = {
+  useOverlay: 'Loading',
+  rowData: [],
+  highlighted: [],
+  putRowIds: [],
+  undoStack: [],
+  summary: [],
+  errors: [],
+  fullRefresh: false,
+
+  // needed by the panels
+  pushUndo: eh.pushUndo,
+  popUndo: eh.popUndo,
+
+  // paste operations must be done row-by-row. build up |pendingPasteUndo|
+  // while in paste mode, then when onPasteEnd is called then call pushUndo
+  pendingPasteUndo: [],
+  pasteMode: false
+};
+
+const defaultSideBar = {
+  toolPanels: [
+    {
+      id: 'findReplace',
+      labelDefault: 'Find Replace',
+      labelKey: 'findReplace',
+      iconKey: 'columns',
+      toolPanel: 'findReplacePanel'
+    }
+  ],
+  defaultToolPanel: ''
+};
+
+const IngestState = Object.freeze({ Loading: 0, Edited: 1, Locked: 2, Valid: 3, ConfirmSubmit: 4 });
+
+interface JobType {
+  program?: string,
+  reference?: string,
+  isExtendedSize?: boolean,
+  source?: SourceJobType,
+  status?: StatusJobType,
+}
+
+interface DataSheetViewProps {
+  onIngest: (resp: AxiosResponse) => void,
```

Comment:
i guess it is a typo here? `resp` maybe `res`?

Reply:
fixed